meeshkan/server/commands.py

```diff
@@ -1,91 +1,114 @@
-@click.option("-m", "--mode", type=click.Choice(['GEN', 'REPLAY', 'MIXED'], case_sensitive=False),
-              default=None, help="Spec building mode.")
-def record(port, admin_port, log_dir, header_routing, specs_dir, mode):
-    """
-    Record http traffic to http-types format.
-    """
-    start_admin(admin_port)
-    logger.info('Starting Meeshkan proxy on http://localhost:%s', port)
-    logger.info('Spec generation mode is %s', mode.lower() if mode else 'disabled')
-    with RequestLoggingCallback(log_dir=log_dir, specs_dir=specs_dir,
-                                update_mode=UpdateMode[mode.upper()] if mode else None) as callback:
-        server = RecordProxy(callback, HeaderRouting() if header_routing else PathRouting())
-        server.listen(port)
-        tornado.ioloop.IOLoop.instance().start()
+    return _add_options
 
 
-class MeeshkanApplication(Application):
-    response_matcher: ResponseMatcher
-    router: Routing
+_common_server_options = [
+    click.option('-p', '--port', default="8000", help='Server port.'),
+    click.option('-a', '--admin_port', default="8888", help='Admin server port.'),
+    click.option('-s', '--specs_dir', default="./specs", help='Directory with OpenAPI schemas.'),
+    click.option('-d', '--daemon', is_flag=True, help='Whether to run meeshkan as a daemon.')
+]
 
 
-def make_mocking_app(callback_path, specs_dir, router):
-    app = MeeshkanApplication([
-        (r'/.*', MockServerView)
-    ])
-    callback_manager.load(callback_path)
+@click.group()
+def mock():
+    pass
 
-    app.response_matcher = ResponseMatcher(specs_dir)
-    app.router = router
-    return app
 
-
-@click.command()
+@mock.command(name='start')
 @click.option('-c', '--callback_path', default="./callbacks", help='Directory with configured callbacks.')
-@click.option('-p', '--port', default="8000", help='Server port.')
-@click.option('-a', '--admin_port', default="8888", help='Admin server port.')
-@click.option('-s', '--specs_dir', default="./specs", help='Directory with OpenAPI schemas.')
 @click.option('-r', '--header_routing', is_flag=True, help='Whether to use a path based routing to a target host.')
-def mock(callback_path, admin_port, port, specs_dir, header_routing):
+@add_options(_common_server_options)
+def start_mock(callback_path, admin_port, port, specs_dir, header_routing, daemon):
     """
     Run a mock server.
     """
-    start_admin(admin_port)
-    app = make_mocking_app(callback_path, specs_dir, HeaderRouting() if header_routing else PathRouting())
-    http_server = HTTPServer(app)
-    http_server.listen(port)
-    logger.info('Mock server is listening on http://localhost:%s', port)
-    IOLoop.current().start()
+    server = MockServer(callback_path=callback_path, admin_port=admin_port, port=port, specs_dir=specs_dir,
+                        routing=HeaderRouting() if header_routing else PathRouting())
+
+    if daemon:
+        daemon = daemonocle.Daemon(
```

Comment:
This is an interesting project! Did it seem solid to you? It hasn't been updated in four years so maybe it's perfect. We can of course make our own fork if it doesn't work as expected.

Reply:
There're a lot of them. This is the most common one probably - https://pypi.org/project/python-daemon/
There's even PEP for this feature: https://www.python.org/dev/peps/pep-3143/
daemonocle is easier to customize. python-daemon can be used only to build simple start/stop/status behaviour.
They're pretty similar under the hood. And the code is quite simple actually.